File: Python/S_Q_DQ/array_queue.py
# author - newguo@sonaspy.cn
# coding - utf_8
import sys
sys.path.append("../")
from myExceptions import Empty


class ArrayQueue:
    """FIFO queue implementation using a Python list as underlying storage."""
    DEFAULT_CAPACITY = 10          # moderate capacity for all new queues

    def __init__(self):
        """Create an empty queue."""
        self._data = [None] * ArrayQueue.DEFAULT_CAPACITY
        self._rear = 0
        self._front = 0

    def __len__(self):
        """Return the number of elements in the queue."""
        return self._rear

    def is_empty(self):
        """Return True if the queue is empty."""
        return self._rear == 0

    def first(self):
        """Return (but do not remove) the element at the front of the queue.
        Raise Empty exception if the queue is empty.
        """
        if self.is_empty():
            raise Empty('Queue is empty')
        return self._data[self._front]

    def dequeue(self):
        """Remove and return the first element of the queue (i.e., FIFO).
        Raise Empty exception if the queue is empty.
        """
        if self.is_empty():
            raise Empty('Queue is empty')
        answer = self._data[self._front]
        self._data[self._front] = None         # help garbage collection
        self._front = (self._front + 1) % len(self._data)
        self._rear -= 1
        return answer

    def enqueue(self, e):
        """Add an element to the back of queue."""
        if self._rear == len(self._data):
            self._resize(2 * len(self.data))     # double the array size
        avail = (self._front + self._rear) % len(self._data)
        self._data[avail] = e
        self._rear += 1

    def _resize(self, cap):                  # we assume cap >= len(self)
        """Resize to a new list of capacity >= len(self)."""
        old = self._data                       # keep track of existing list
        # allocate list with new capacity
        self._data = [None] * cap
        walk = self._front
        for k in range(self._rear):            # only consider existing elements
            self._data[k] = old[walk]            # intentionally shift indices
            walk = (1 + walk) % len(old)         # use old size as modulus
        self._front = 0                        # front has been realigned

File: Cpp/arrayQueue.cpp
// test array queue

#include <iostream>
#include "arrayQueue.h"
#include "myExceptions.h"

using namespace std;

int main(void)
{
   arrayQueue<int> q(4);

   // add a few elements
   q.push(1);
   std::cout << "Queue rear is " << q.back() << std::endl;
   q.push(2);
   std::cout << "Queue rear is " << q.back() << std::endl;
   q.push(3);
   std::cout << "Queue rear is " << q.back() << std::endl;
   q.push(4);
   std::cout << "Queue rear is " << q.back() << std::endl;

   std::cout << "Queue should be 1234, front to rear" << std::endl;

   // test empty and size
   if (q.empty())
      std::cout << "The queue is empty" << std::endl;
   else
      std::cout << "The queue is not empty" << std::endl;

   std::cout << "The queue size is " << q.size() << std::endl;

   while (!q.empty())
   {
      std::cout << "Queue front is " << q.front() << std::endl;
      q.pop();
      std::cout << "Popped front element" << std::endl;
   }

   try {q.pop();}
   catch (queueEmpty message)
   {
      std::cout << "Last pop failed " << std::endl;
      message.outputMessage();
   }

   // create a wraparound queue and do array doubling
   arrayQueue<int> r(4);
   r.push(1);
   r.push(2);
   r.push(3);
   r.pop();
   r.pop();
   r.push(4);
   r.push(5);
   r.push(6);
   r.push(7);

   std::cout << "Queue should be 34567, front to rear" << std::endl;

   // test empty and size
   if (r.empty())
      std::cout << "The queue is empty" << std::endl;
   else
      std::cout << "The queue is not empty" << std::endl;

   std::cout << "The queue size is " << r.size() << std::endl;

   while (!r.empty())
   {
      std::cout << "Queue front is " << r.front() << std::endl;
      r.pop();
      std::cout << "Popped front element" << std::endl;
   }

   return 0;
}
